### ISSUE TITLE
BLIP: Unable to add todo users after deleting the user

### ISSUE THREAD

**Issue by raghusarangapani** [couchbase/sync_gateway#2718]
### Sync Gateway version
1.5.0-474 with CBL 2.0 blip replication

### Operating system
CentOS 7.

### Config file
https://github.com/couchbaselabs/mobile-testkit/blob/master/resources/sync_gateway_configs/testfest_todo_cc.json

### Log output
```
2017-07-14T20:46:27.195Z WARNING: Sync fn exception: TypeError; doc = map[_id:raghu.6e46aadb-19d9-4bf2-9f57-40b4db3670b3.sridevi _rev:2-9d5b1f57d8e607d54f6625f07b42c1edcc821256 _deleted:%!s(bool=true)] -- db.(*Database).getChannelsAndAccess() at crud.go:1158
2017-07-14T20:46:27.195Z CRUD+: Did not update document "raghu.6e46aadb-19d9-4bf2-9f57-40b4db3670b3.sridevi" w/ xattr: 500 Exception in JS sync function
```
Full log:
https://gist.github.com/raghusarangapani/aaa6fb3bfe398e5f1c35416ff1548474

### Expected behavior
User should be able to add another user

### Actual behavior
Unable to add users after deleting the user

### Steps to reproduce
1.  Two users login to todo app on two devices.
2. Create lists, add tasks and share it with other users.
3. Remove the users from the sharing list for the lists and then re-share with the same user again.

**Comment by tleyden** [comment]
That looks like it's potentially a bug in the sync function, or the data itself.  Was this sync function (unmodified) working against previous versions of Sync Gateway?

**Comment by raghusarangapani** [comment]
Yes. The only change to sync gateway config is to enable xattrs, blip and autoimport. Rest of the config was not changed.

**Comment by tleyden** [comment]
Although, it's a whole different client library being tested.

To help isolate it, it would be valuable to run this with Xattrs disabled.  I don't think autoimport is a factor here, since presumably nothing is writing to the couchbase server bucket directly (though that would really make an awesome test fest!)

**Comment by tleyden** [comment]
Also, it would be great to capture this into a functional test.  A network capture might help.  Oh wait, we don't have a way to interpret blip traffic, so that won't work.

**Comment by adamcfraser** [comment]
Looking at the logs, there are a number of documents being rejected by the Sync Function with Type Errors (not just task list user docs).

Most of these docs appear to be conflicting tombstone revisions (e.g. rev 2-abc is successfully written, then rev 2-def _deleted=true hits the error).  

I expect the sync function error are related to the oldDoc value that's being passed to the sync function - in the case of these conflicting tombstone revisions, oldDoc would be nil.  From that perspective it's a sync function issue - when using oldDoc, it should be doing an existence check first.

However, I think there's an underlying problem that either 
(a) CBL isn't including the revision history when sending the tombstone, and so that's causing oldDoc to be nil, or 
(b) Sync Gateway isn't correctly identifying oldDoc for a conflicting tombstone revision.

Before digging into this further, it would be useful to repro this issue using the latest CBL and SG, to ensure this isn't something that's been fixed since this was filed.  @raghusarangapani Do we have any upcoming test fests planned for CBL 2.0 + SG 1.5?

**Comment by raghusarangapani** [comment]
Not with CBL 2.0, I plan on having a test fest with SG 1.5 + CBL 1.4.

**Comment by adamcfraser** [comment]
Ok - moving this out of the current sprint to wait until we can repro w/ 2.0

**Comment by tleyden** [comment]
Just attempted to repro against SG 

9f4a7bd7ba951c7d8469413e5ea60b124879138b (latest commit on `feature/issue_3136_blip3`), and was not able to reproduce the behavior or any TypeErrors in the SG logs.

I did notice an issue where removing a user from a share didn't remove those items from that other users list.

On the left is user2, and the right is user1

![screen shot 2018-01-10 at 2 22 26 pm](https://user-images.githubusercontent.com/296876/34798763-2a1118ec-f612-11e7-9297-4859fd2a6d08.png)


after removing user1 from the list of authorized users, I would expect the HelloFromUser2 list to disappear, but it didn't.  (even after restarting todo)

![screen shot 2018-01-10 at 2 24 09 pm](https://user-images.githubusercontent.com/296876/34798770-2f85c426-f612-11e7-996b-cf6edf2b44d3.png)


Closing this out since I don't see any obvious SG bugs.
